### DISCUSSION TITLE
sqlalchemy.ext.asyncio can't use transcation, when connect mysql

### DISCUSSION THREAD

**Discussion by lishu2006ll** [sqlalchemy/sqlalchemy#11876] [Usage Questions]
### Describe the bug

In an asynchronous environment, many methods have been tried but the transaction rollback does not work, but in the synchronous method it works.

### SQLAlchemy Version in Use

2.0.32

### DBAPI (i.e. the database driver)

aiomysql 0.2.0

### Database Vendor and Major Version

mysql 5.7.36

### Python Version

3.8

### Operating system

linux, osx

### To Reproduce

```python
self.db is :AsyncSession
async with self.db.begin():  # 开始一个新的事务
            try:
                if obj_in.parent_id != 0:
                    await self.get(id=obj_in.parent_id)
                new_obj = await self.create(obj_in=obj_in)
                await self.db.commit()
                raise Exception("xxxxx")
            except Exception as e:
                print(e)
                await self.db.rollback()


or :
    try:
                if obj_in.parent_id != 0:
                    await self.get(id=obj_in.parent_id)
                new_obj = await self.create(obj_in=obj_in)
                await self.db.commit()
                raise Exception("xxxxx")
            except Exception as e:
                print(e)
                await self.db.rollback()
```

### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
``` 
no error, but the db table have a record, can't rollback()

**Comment by CaselIT** [comment]
Hi,

There is no reproduction code and it's missing a lot of information, so this is a discussion.

What is db? How it's created?
Can you post a runnable example that your issue?
What do the show when you set echo=True in create_engine?

**Reply by lishu2006ll** [reply]
> Hi,
> 
> There is no reproduction code and it's missing a lot of information, so this is a discussion.
> 
> What is db? How it's created? Can you post a runnable example that your issue? What do the show when you set echo=True in create_engine?
I use mysql 5
the code is :
   the connect code is :
![image](https://github.com/user-attachments/assets/a77aafc7-848d-4560-b481-7105917aadb8)

    async def create_dept(self, obj_in: DeptCreate):
        # 创建
        async with self.db.begin():  # 开始一个新的事务
            try:
                #if obj_in.parent_id != 0:
                #    await self.get(id=obj_in.parent_id)
                new_obj = await self.create(obj_in=obj_in)
                await self.db.commit()
                raise Exception("xxxxx")
            except Exception as e:
                print("rollback")
                await self.db.rollback()

the execute logs is: 
![Uploading image.png…]()

**Reply by lishu2006ll** [reply]
![image](https://github.com/user-attachments/assets/0f5808fe-a4f8-424d-9d24-19f1d07a3494)

**Reply by CaselIT** [reply]
How do you create db?

**Reply by lishu2006ll** [reply]
as fllower: i use the create_async_session when i will use self.db
def create_async_session(self):  
        return self.sessionmaker() 

db = async_db_inst.create_async_session()

**Reply by lishu2006ll** [reply]
# 创建一个会话工厂  
                self.sessionmaker = async_sessionmaker(  
                    bind=self.engine,  
                    autoflush=False,  
                    expire_on_commit=False  
                )

**Comment by zzzeek** [comment]
when you use a with: block, the rollback and commit are a given - change:

```
async with self.db.begin():  # 开始一个新的事务
            try:
                if obj_in.parent_id != 0:
                    await self.get(id=obj_in.parent_id)
                new_obj = await self.create(obj_in=obj_in)
                await self.db.commit()
                raise Exception("xxxxx")
            except Exception as e:
                print(e)
                await self.db.rollback()
```

to

```
async with self.db.begin():  # 开始一个新的事务
            if obj_in.parent_id != 0:
               await self.get(id=obj_in.parent_id)
            new_obj = await self.create(obj_in=obj_in)
```

additionally, *never* swallow the exception in an "except:" block:


```
try:
    # ...
except Exception as e:
   print(e)

   raise e   # <---- ALWAYS do this!!!
```

if you are seeing errors that commit or rollback is not possible, it's likely because the "with" block is attempting to commit the transaction (since you are swallowing the exception) but you've already rolled it back.
